### ISSUE TITLE
Commits and pull files

### ISSUE THREAD

**Issue by lunny** [go-gitea/gitea#124]
Before we can start a review, we can implement comment on one line of PRs at first.
I think this includes two sub issues. 

* [ ] ~Commit comment #3748~ see #4898
* [x] Pull files comment #3748

**Comment by bkcsoft** [comment]
well, If we implement `comments on commits` we should get this "for free" :slightly_smiling_face:

**Comment by metalmatze** [comment]
I'm not sure. I don't know the internals, but the comment may go onto a commit page. Then you can discuss this commit. 
But this issue targets a bigger picture. A PR might have multiple commits and those can change over time. So I agree that comments on commits might benefit comments on PR with some work already done nevertheless I feel that there's more to it.

**Comment by lunny** [comment]
Yes, comment on commits.

**Comment by bkcsoft** [comment]
I can probably make a proof of concept for this sometime during the weekend or next week

**Comment by lunny** [comment]
@bkcsoft any update?

**Comment by strk** [comment]
Aren't we going to do this with git-appraise ? See #733

**Comment by sokolovstas** [comment]
I can look at this, but this will be not so fast

**Comment by lunny** [comment]
Move this v1.2 since #733 has moved.

**Comment by lunny** [comment]
I have tried to do something. `Comments on PR` is not the same as `comments on commits`. These are two similar PRs. For `comments on PR`, all the comments should be indexed by the IssueID, treePath, lineNum and actionType(add line or del line). But for `comments on commits`,  all the comments should be indexed by commitSHA, treePath, lineNum and actionType(add line or del line).

**Comment by lcges** [comment]
I send the draft as seen to comment on the code.
For some time I will show demo work.
![1](https://user-images.githubusercontent.com/19255071/30315213-752c5512-97a3-11e7-9b75-f6d1679d5a99.png)

**Comment by lcges** [comment]
Hello.
I send you a preview of the model of action.
This is a static implementation - proof of concept

[gitea_comment_code.zip](https://github.com/go-gitea/gitea/files/1324115/gitea.zip)

**Comment by daviian** [comment]
@qdbdbp Looks great but I think a button to cancel the comment is missing. How does it look in unified mode?

**Comment by lcges** [comment]
[gitea_c_o_c.zip](https://github.com/go-gitea/gitea/files/1324606/gitea.1.zip)

I send updated versions.
I'm waiting for comments on this.

![pasted image at 2017_09_22 03_04 pm](https://user-images.githubusercontent.com/19255071/30745729-973e905a-9fa7-11e7-8927-5ceafb2f8353.png)

**Comment by lunny** [comment]
@qdbdbp I am working on a branch about this, I will send a PR today or last two days. I will copy your code to that PR.

**Comment by lunny** [comment]
@qdbdbp It seems only the first comment on the code line. how to show the second or third comment on the same line?

**Comment by lunny** [comment]
@qdbdbp I have sent a PR #2583 but I think maybe I have lost something from your zip. Please review the code and help me with the UI.

**Comment by lcges** [comment]
no problem. I will do what I need
I'll take your code and I'm already working on the correct code.

![pasted image at 2017_09_25 08_30 am](https://user-images.githubusercontent.com/19255071/30795258-f8360e16-a1cc-11e7-8a93-3a4cb514af2f.png)
![pasted image at 2017_09_25 08_35 am](https://user-images.githubusercontent.com/19255071/30795259-f8360088-a1cc-11e7-900d-aae3cf83f63b.png)

**Comment by tonivj5** [comment]
@qdbdbp, what tool do you use to do those designs? :sweat_smile:

**Comment by lcges** [comment]
hi. I use it for quick drafts this tool:

https://mockflow.com/

**Comment by bclermont** [comment]
I don't mind to pay to have this feature... how much? who?

**Comment by lunny** [comment]
@bclermont we already have a wip pull request #2583.

**Comment by bclermont** [comment]
@lunny yeah, but it can't be merged, and haven't been updated in 6 months! I want to help to implement this feature

**Comment by lunny** [comment]
You could use https://www.bountysource.com/teams/go-gitea

**Comment by techknowlogick** [comment]
@lunny thanks for posting that link. I added to the bounty.

**Comment by cobbzilla** [comment]
Thanks for the bounty link. I added $100 if this can be resolved within 6 months. This is the basically the *only* feature that I still miss after migrating from Github to Gitea. I will be super duper happy when Gitea supports in-line code reviewing for pull requests. Go Gitea team!

**Comment by bclermont** [comment]
I will give equivalent of 100$ in a crypto currency to whoever earn the bounty

**Comment by rc1021** [comment]
Are there things to follow?

**Comment by lunny** [comment]
Please follow #3748

**Comment by tonivj5** [comment]
Is comment in commits implemented? If it isn't implemented I think this should be reopen

**Comment by lunny** [comment]
@xxxtonixxx right, Only pull request comment was implemented by #3748 . @lafriks @JonasFranzDEV

**Comment by lunny** [comment]
But we could open another issue to keep this one closed.

**Comment by tonivj5** [comment]
:+1:, I will do it

**Comment by lunny** [comment]
@xxxtonixxx Thank you!
